File: Leetcode/2-AddTwoNumbers/2-AddTwoNumbers.cpp
// Last updated: 6/7/2026, 11:28:05 PM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* cur1 = l1;
        ListNode* cur2 = l2;
        ListNode* head = nullptr;
        ListNode* prev = nullptr;
        int place=0;
        int carry=0;
        while(cur1!=nullptr&&cur2!=nullptr)
        {
            place = cur1->val+cur2->val+carry;
            carry = place/10;
            place = place%10;
            ListNode* cur = new ListNode(place);
            if(prev==nullptr)
            {
                head = cur;
            }
            else
            {
                prev->next=cur;
            }
            prev = cur;
            cur1=cur1->next;
            cur2=cur2->next;
        }
        if(cur2!=nullptr)
        {
            swap(cur1,cur2);
        }
        while(cur1!=nullptr)
        {
            place = cur1->val+carry;
            carry = place/10;
            place=place%10;
            ListNode* cur = new ListNode(place);
            if(prev==nullptr)
            {
                head=cur;
            }
            else
            {
                prev->next=cur;
            }
            prev=cur;
            cur1=cur1->next;
        }
        if(carry)
        {
            ListNode* cur = new ListNode(carry);
            prev->next = cur;
        }
        return head;
    }
};

File: Leetcode/2-AddTwoNumbers/2-AddTwoNumbers.java
// Last updated: 6/7/2026, 11:28:08 PM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) 
    {
        ListNode ans = new ListNode();
        int carry = 0;
        int add =0;
        int put = 0;
        ListNode nav = ans;
        while(l1!=null||l2!=null)
        {
            while(l1!=null&&l2!=null)
            {
                ListNode n = new ListNode();
                add=l1.val+l2.val+carry;
                put=add%10;
                carry=add/10;
                n.val=put;
                l1=l1.next;
                l2=l2.next;
                nav.next=n;;
                nav=nav.next;
            }
            while(l1!=null)
            {
                ListNode n = new ListNode();
                add=l1.val+carry;
                put=add%10;
                carry=add/10;
                n.val=put;
                l1=l1.next;
                nav.next=n;
                nav=nav.next;
            }
            while(l2!=null)
            {
                ListNode n = new ListNode();
                add=l2.val+carry;
                put=add%10;
                carry=add/10;
                n.val=put;
                l2=l2.next;
                nav.next=n;
                nav=nav.next;
            }
            if(carry==1)
            {
                ListNode n = new ListNode();
                n.val=carry;
                nav.next=n;
                nav=nav.next;
            }
        }
        ans=ans.next;
        return ans;
    }
}